File: app/src/main/java/net/ginteam/carmen/presenter/auth/SignUpPresenter.java
package net.ginteam.carmen.presenter.auth;

import android.util.Log;

import net.ginteam.carmen.contract.auth.SignUpContract;
import net.ginteam.carmen.model.auth.UserModel;
import net.ginteam.carmen.provider.ModelCallback;
import net.ginteam.carmen.provider.auth.AuthProvider;

/**
 * Created by Eugene on 1/4/17.
 */

public class SignUpPresenter implements SignUpContract.Presenter {

    private SignUpContract.View mView;

    @Override
    public void signUp(String name, String email, String password) {
        mView.showLoading(true);

        AuthProvider
                .getInstance()
                .userRegister(name, email, password, new ModelCallback<UserModel>() {
                    @Override
                    public void onSuccess(UserModel resultModel) {
                        mView.showLoading(false);
                        mView.showMain();
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showLoading(false);
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void attachView(SignUpContract.View view) {
        mView = view;
    }

    @Override
    public void detachView() {
        mView = null;
    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/presenter/authentication/SignUpPresenter.kt
package net.ginteam.carmen.kotlin.presenter.authentication

import net.ginteam.carmen.R
import net.ginteam.carmen.kotlin.api.response.ModelSubscriber
import net.ginteam.carmen.kotlin.contract.SignUpContract
import net.ginteam.carmen.kotlin.model.AuthModel
import net.ginteam.carmen.kotlin.presenter.BasePresenter
import net.ginteam.carmen.kotlin.provider.AuthProvider
import net.ginteam.carmen.kotlin.provider.AuthenticationProvider

/**
 * Created by eugene_shcherbinock on 2/15/17.
 */
class SignUpPresenter : BasePresenter <SignUpContract.View>(), SignUpContract.Presenter {

    private val mAuthProvider: AuthProvider = AuthenticationProvider

    override fun signUp(name: String, email: String, password: String) {
        mView?.showLoading(true, R.string.registration_progress_message)

        mAuthProvider
                .userRegister(name, email, password)
                .subscribe(object : ModelSubscriber<AuthModel>() {
                    override fun success(model: AuthModel) {
                        mView?.showLoading(false)
                        mView?.showMainActivity()
                    }

                    override fun error(message: String, isNetworkError: Boolean) {
                        mView?.showError(message)
                    }
                })
    }
}